src/util/alloc/allocator.rs

```diff
@@ -102,32 +116,70 @@ pub fn get_maximum_aligned_size<VM: VMBinding>(
-    /// For example, bump pointer allocator acquire memory at 32KB blocks. Depending on the actual size for the current object,
-    /// they always acquire memory of N*32KB (N>=1). Thus bump pointer allocator returns 32KB for this method.
-    /// Only allocators that do thread local allocation need to implement this method.
+    /// Return at which granularity the allocator acquires memory from the global space and use
+    /// them as thread local buffer. For example, the [`BumpAllocator`] acquires memory at 32KB
+    /// blocks. Depending on the actual size for the current object, they always acquire memory of
+    /// N*32KB (N>=1). Thus the [`BumpAllocator`] returns 32KB for this method.  Only allocators
+    /// that do thread local allocation need to implement this method.
     fn get_thread_local_buffer_granularity(&self) -> usize {
         assert!(self.does_thread_local_allocation(), "An allocator that does not thread local allocation does not have a buffer granularity.");
         unimplemented!()
     }
 
+    /// An allocation attempt. The implementation of this function depends on the allocator used.
+    /// If an allocator supports thread local allocations, then the allocation will be serviced
+    /// from its TLAB, otherwise it will default to using the slowpath, i.e. [`alloc_slow`].
+    ///
+    /// Note that in the case where the VM is out of memory, we return a null pointer back to the
+    /// VM and expect that the OOM error is thrown by the binding.
+    ///
+    /// Arguments:
+    /// * `size`: the allocation size in bytes.
+    /// * `align`: the required alignment in bytes.
+    /// * `offset` the required offset in bytes.
     fn alloc(&mut self, size: usize, align: usize, offset: isize) -> Address;
 
+    /// Slowpath allocation attempt. We break up the slowpath implementation into two parts
+    /// in order to improve efficiency of our code by reducing cache misses.
+    ///
+    /// Arguments:
+    /// * `size`: the allocation size in bytes.
+    /// * `align`: the required alignment in bytes.
+    /// * `offset` the required offset in bytes.
     #[inline(never)]
     fn alloc_slow(&mut self, size: usize, align: usize, offset: isize) -> Address {
         self.alloc_slow_inline(size, align, offset)
     }
 
+    /// Slowpath allocation attempt. This function executes the actual slowpath allocation.  A
+    /// slowpath allocation in MMTk attempts to allocate the object using the per-allocator
+    /// definition of [`alloc_slow_once`]. This function also accounts for increasing the
+    /// allocation bytes in order to support stress testing. In case precise stress testing is
+    /// being used, the [`alloc_slow_once_precise_stress`] function is used instead.
+    ///
+    /// Note that in the case where the VM is out of memory, we throw a [`HeapOutOfMemory`] error
+    /// and return a null pointer back to the VM with the expectation that the OOM error is thrown
```

Comment:
I would suggest that we do not make any assumption about how a binding would implement `out_of_memory()` -- they could just panic, or return and throw their exception later. It is the binding's choice, and we do not need to make any assumption on that.

We only need to make it clear that in case of OOM we will invoke `out_of_memory()`, and then the method will return null pointer.

Reply:
I agree. I thought that's essentially what I said, but "expectation" is too strong. I will update the comments.